pkg/library/ssh/event.go

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2019-2024 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ssh
+
+import (
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/devworkspace-operator/pkg/constants"
+	"github.com/devfile/devworkspace-operator/pkg/library/lifecycle"
+)
+
+const commandLine = `SSH_ENV_PATH=$HOME/ssh-environment \
+&& if [ -f /etc/ssh/passphrase ] && command -v ssh-add >/dev/null; \
+then ssh-agent | sed 's/^echo/#echo/' > $SSH_ENV_PATH \
+&& chmod 600 $SSH_ENV_PATH && source $SSH_ENV_PATH \
+&& ssh-add /etc/ssh/dwo_ssh_key < /etc/ssh/passphrase \
+&& if [ -f $HOME/.bashrc ]; then echo "source ${SSH_ENV_PATH}" >> $HOME/.bashrc; fi; fi`
```

Comment:
Modifying the `.bashrc` works, but if a user provides their own custom `.bashrc` through an [automount configmap](https://github.com/devfile/devworkspace-operator/blob/main/docs/additional-configuration.adoc#automatically-mounting-volumes-configmaps-and-secrets), this feature will actually break. Files that are mounted via the automount configmap mechanism cannot be written to.

Instead, we could place the SSH_ENV_PATH in `/etc/profile.d/`, which contains scripts that should be run when sourcing the `.bashrc`, e.g. `SSH_ENV_PATH=/etc/profile.d/.ssh-environment`. This way, any `.bashrc` will load the ssh-agent script.

```bash
$ cat ~/.bashrc
# .bashrc

# Source global definitions
if [ -f /etc/bashrc ]; then
    . /etc/bashrc
fi
(...)


$ cat /etc/bashrc 
(...)
    SHELL=/bin/bash
    # Only display echos from profile.d scripts if we are no login shell
    # and interactive - otherwise just process them to set envvars
    for i in /etc/profile.d/*.sh; do
        if [ -r "$i" ]; then
            if [ "$PS1" ]; then
                . "$i" # Here is where all the scripts in /etc/profile.d/ get sourced
            else
                . "$i" >/dev/null
            fi
        fi
    done
```

A caveat is that the /etc/profile.d/ scripts only get run for interactive shells. But for https://github.com/devfile/devworkspace-operator/issues/1295 this is fine.

**I do understand that this is a rather big change to request from your PR @vinokurig, so I'm perfectly fine with keeping the current approach that modifies the .bashrc and making a new issue/PR in the future to implement my suggestions.** I haven't actually tested my suggestion, so it might not actually work (though I did [something similar](https://github.com/devfile/developer-images/pull/115/commits/7ac79710c16454500116b51c37e3bdffc30e66de) for the UDI a while back)

Reply:
> I do understand that this is a rather big change to request from your PR @vinokurig, so I'm perfectly fine with keeping the current approach that modifies the .bashrc and making a new issue/PR in the future to implement my suggestions. 

I think we should take the current approach for now since there isn't a straight forward alternative since the `/etc/profile.d` folder cannot be written by regular users by default in the UDI. I suggest we keep the functionality how it is for now, and document that if .bashrc is being mounted using a configmap, something like `[ -f /home/user/ssh-environment ] && source /home/user/ssh-environment` should be added to it.

Any thoughts?

Reply:
Good catch :) I agree that it's probably worth documenting this somewhere in the near future (maybe in the Che Docs as well as DWO additional docs)?